File: app/src/test/java/vassar/cmpu203/vassardiningapp/UserTest.java
package vassar.cmpu203.vassardiningapp;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import vassar.cmpu203.vassardiningapp.model.DietaryRestriction;
import vassar.cmpu203.vassardiningapp.model.MealtimeItem;
import vassar.cmpu203.vassardiningapp.model.MealtimeMenu;
import vassar.cmpu203.vassardiningapp.model.User;

public class UserTest {
    User user;
    Set<DietaryRestriction> restrictions;
    MealtimeItem testItem;

    UserTest() {
        user = new User();
        restrictions = new HashSet<>(List.of(DietaryRestriction.VEGETARIAN, DietaryRestriction.KOSHER));
        user.setDietaryRestrictions(restrictions);
        testItem = new MealtimeItem("test", "123","test desc", "", restrictions);
    }

    @Test
    public void testSwitchStatus() {
        user.switchFavoriteStatus(testItem);
        assertTrue(user.getFavorites().containsValue(testItem));
        user.switchFavoriteStatus(testItem);
        assertFalse(user.getFavorites().containsValue(testItem));
        assertTrue(user.getDietaryRestrictions().contains(DietaryRestriction.VEGETARIAN));
        user.switchRestrictionStatus(DietaryRestriction.VEGETARIAN);
        assertFalse(user.getDietaryRestrictions().contains(DietaryRestriction.VEGETARIAN));
    }

    @Test
    public void testMatchRestriction() {
        assertTrue(user.matchRestriction(restrictions));
        assertTrue(user.matchRestriction(new HashSet<>(List.of(DietaryRestriction.VEGAN))));
        assertFalse(user.matchRestriction(new HashSet<>(List.of(DietaryRestriction.KOSHER))));
        assertFalse(user.matchRestriction(new HashSet<>(List.of(DietaryRestriction.GLUTEN_FREE, DietaryRestriction.KOSHER))));
    }

    @Test
    public void testFilterMenus() {
        List<MealtimeMenu> menus = new ArrayList<>();
        menus.add(new MealtimeMenu("cafe", "date", "mealtime", new ArrayList<>(List.of(testItem))));
        user.toggleFavoriteFilter();
        assertFalse(user.filterMenus(menus).get(0).getMenuItems().contains(testItem));
        user.switchFavoriteStatus(testItem);
        assertTrue(user.filterMenus(menus).get(0).getMenuItems().contains(testItem));
    }
}

File: app/src/test/java/vassar/cmpu203/vassardiningapp/UserTest.kt
package vassar.cmpu203.vassardiningapp

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import vassar.cmpu203.vassardiningapp.model.DietaryRestriction
import vassar.cmpu203.vassardiningapp.model.MealtimeItem
import vassar.cmpu203.vassardiningapp.model.MealtimeMenu
import vassar.cmpu203.vassardiningapp.model.User

class UserTest internal constructor() {
    private var user: User = User()
    private var testItem: MealtimeItem
    private var restrictions: MutableSet<DietaryRestriction> =
            HashSet(listOf(DietaryRestriction.VEGETARIAN, DietaryRestriction.KOSHER))

    init {
        user.dietaryRestrictions = restrictions
        testItem = MealtimeItem("test", "123", "test desc", "", restrictions)
    }

    @Test
    fun testSwitchStatus() {
        user.switchFavoriteStatus(testItem)
        Assertions.assertTrue(user.favorites.containsValue(testItem))
        user.switchFavoriteStatus(testItem)
        Assertions.assertFalse(user.favorites.containsValue(testItem))
        Assertions.assertTrue(user.dietaryRestrictions.contains(DietaryRestriction.VEGETARIAN))
        user.switchRestrictionStatus(DietaryRestriction.VEGETARIAN)
        Assertions.assertFalse(user.dietaryRestrictions.contains(DietaryRestriction.VEGETARIAN))
    }

    @Test
    fun testMatchRestriction() {
        Assertions.assertTrue(user.matchRestriction(restrictions))
        Assertions.assertTrue(user.matchRestriction(HashSet(listOf(DietaryRestriction.VEGAN))))
        Assertions.assertFalse(user.matchRestriction(HashSet(listOf(DietaryRestriction.KOSHER))))
        Assertions.assertFalse(user.matchRestriction(HashSet(listOf(DietaryRestriction.GLUTEN_FREE, DietaryRestriction.KOSHER))))
    }

    @Test
    fun testFilterMenus() {
        val menus: MutableList<MealtimeMenu> = ArrayList()
        menus.add(MealtimeMenu("cafe", "date", "mealtime", ArrayList(listOf(testItem))))
        user.toggleFavoriteFilter()
        Assertions.assertFalse(user.filterMenus(menus)[0].menuItems.contains(testItem))
        user.switchFavoriteStatus(testItem)
        Assertions.assertTrue(user.filterMenus(menus)[0].menuItems.contains(testItem))
    }
}